File: checkdigit/workspace/stream.py
"""
stream.py
=========
Chunked readers for large files. Bytes are read in fixed chunks and decoded
with an incremental decoder so multibyte sequences split across a chunk
boundary are never corrupted. Records are yielded with the absolute character
offset (in the decoded text) of their raw span, so a later pass can splice
edits in place while streaming the same file again.

Readers:
  * text_chunks      : bytes -> decoded text chunks with a chosen codec
  * csv_records      : RFC 4180 records across quoted newlines and chunk boundaries
  * line_records     : one record per physical line (plain text, fixed-width)
  * edifact_segments : segments across escaped delimiters and chunk boundaries

The codec is chosen once from the first chunk: strict UTF-8 if it decodes,
otherwise ISO-8859-1 for the whole file (the same rule as service.py).
"""
from __future__ import annotations

import codecs
import re
from dataclasses import dataclass
from typing import Iterator, List, Optional, Tuple

CHUNK_BYTES = 1024 * 1024


def choose_codec(head: bytes) -> str:
    """Codec guess from a prefix; only reliable when `head` is the whole file."""
    try:
        codecs.getincrementaldecoder("utf-8")(errors="strict").decode(head, final=True)
        return "utf-8"
    except UnicodeDecodeError:
        return "latin-1"


def sniff_codec(path: str, chunk_bytes: int = CHUNK_BYTES) -> str:
    """Strict UTF-8 over the whole file, else ISO-8859-1. One sequential read."""
    dec = codecs.getincrementaldecoder("utf-8")(errors="strict")
    try:
        with open(path, "rb") as fh:
            while True:
                chunk = fh.read(chunk_bytes)
                if not chunk:
                    dec.decode(b"", final=True)
                    return "utf-8"
                dec.decode(chunk, final=False)
    except UnicodeDecodeError:
        return "latin-1"


def text_chunks(path: str, codec: Optional[str] = None, chunk_bytes: int = CHUNK_BYTES
                ) -> Iterator[Tuple[str, str]]:
    """Yield (codec, text_chunk). Every item carries the codec in use.

    Pass the codec recorded for the source when it is known; otherwise the
    whole file is sniffed first, because a prefix cannot prove UTF-8.
    """
    codec = codec or sniff_codec(path, chunk_bytes)
    with open(path, "rb") as fh:
        head = fh.read(chunk_bytes)
        dec = codecs.getincrementaldecoder(codec)(errors="strict")
        text = dec.decode(head, final=not head)
        if text:
            yield codec, text
        if not head:
            return
        while True:
            chunk = fh.read(chunk_bytes)
            if not chunk:
                tail = dec.decode(b"", final=True)
                if tail:
                    yield codec, tail
                return
            text = dec.decode(chunk, final=False)
            if text:
                yield codec, text


def texts(path: str, codec: Optional[str] = None, chunk_bytes: int = CHUNK_BYTES) -> Iterator[str]:
    for _codec, text in text_chunks(path, codec, chunk_bytes):
        yield text


@dataclass
class Field:
    col: int
    value: str
    start: int        # absolute char offset of the raw field text (including quotes)
    end: int
    quoted: bool


@dataclass
class Record:
    no: int           # 0-based record index (a header row counts)
    start: int        # absolute char offset of the record start
    end: int          # absolute char offset just past the record text (before the line break)
    fields: List[Field]


def csv_records(chunks: Iterator[str], delimiter: str) -> Iterator[Record]:
    """RFC 4180 state machine over a stream of text chunks.

    States: plain field, quoted field, and "after a quote inside a quoted
    field" (the next character decides between an escaped quote and the end
    of the quoted section). The state survives chunk boundaries, as does a
    trailing CR that may be the first half of CRLF. Runs of ordinary
    characters are copied with a regex scan rather than one character at a
    time.
    """
    special = re.compile("[" + re.escape(delimiter) + "\"\r\n]")
    row: List[Field] = []
    buf: List[str] = []
    in_quotes = False
    after_quote = False
    quoted_field = False
    field_start = 0
    record_start = 0
    pos = 0
    rec_no = 0
    pending_cr = False
    saw_any = False

    for chunk in chunks:
        i = 0
        n = len(chunk)
        if n:
            saw_any = True
        if pending_cr:
            pending_cr = False
            if n and chunk[0] == "\n":
                i = 1
                pos += 1
                field_start = pos
                record_start = pos
        while i < n:
            if in_quotes:
                if after_quote:
                    after_quote = False
                    if chunk[i] == '"':
                        buf.append('"')
                        i += 1
                        pos += 1
                        continue
                    in_quotes = False
                    # the character at i is handled below as plain text
                else:
                    j = chunk.find('"', i)
                    if j < 0:
                        buf.append(chunk[i:])
                        pos += n - i
                        i = n
                        continue
                    buf.append(chunk[i:j])
                    pos += j - i + 1
                    i = j + 1
                    after_quote = True
                    continue
            ch = chunk[i]
            if ch == '"' and pos == field_start:
                in_quotes = True
                quoted_field = True
                i += 1
                pos += 1
                continue
            if ch == delimiter:
                row.append(Field(len(row), "".join(buf), field_start, pos, quoted_field))
                buf = []
                quoted_field = False
                i += 1
                pos += 1
                field_start = pos
                continue
            if ch == "\r" or ch == "\n":
                row.append(Field(len(row), "".join(buf), field_start, pos, quoted_field))
                buf = []
                quoted_field = False
                rec = Record(rec_no, record_start, pos, row)
                row = []
                rec_no += 1
                if ch == "\r":
                    if i + 1 < n:
                        step = 2 if chunk[i + 1] == "\n" else 1
                    else:
                        pending_cr = True
                        step = 1
                else:
                    step = 1
                i += step
                pos += step
                field_start = pos
                record_start = pos
                yield rec
                continue
            m = special.search(chunk, i)
            j = m.start() if m else n
            buf.append(chunk[i:j])
            pos += j - i
            i = j
    if saw_any and (buf or row or quoted_field or pos > record_start):
        row.append(Field(len(row), "".join(buf), field_start, pos, quoted_field))
        yield Record(rec_no, record_start, pos, row)


def line_records(chunks: Iterator[str]) -> Iterator[Tuple[int, int, str]]:
    """Yield (line_no_1based, absolute_start, line_text_without_break)."""
    carry = ""
    pos = 0
    line_no = 0
    pending_cr = False
    breaks = re.compile("[\r\n]")
    for chunk in chunks:
        text = carry + chunk
        carry = ""
        start = 0
        n = len(text)
        if pending_cr:
            pending_cr = False
            if n and text[0] == "\n":
                start = 1
        i = start
        while i < n:
            m = breaks.search(text, i)
            if m is None:
                break
            i = m.start()
            line_no += 1
            yield line_no, pos + start, text[start:i]
            if text[i] == "\r":
                if i + 1 < n:
                    i += 2 if text[i + 1] == "\n" else 1
                else:
                    pending_cr = True
                    i += 1
            else:
                i += 1
            start = i
        carry = text[start:]
        pos += start
    if carry:
        line_no += 1
        yield line_no, pos, carry


def edifact_segments(chunks: Iterator[str], release: str, terminator: str) -> Iterator[Tuple[int, str]]:
    """Yield (absolute_start, raw_segment) honouring the release character.

    The raw segment keeps release characters so the caller can splice it back
    verbatim; a UNA header (9 characters) is skipped, not yielded.
    """
    buf: List[str] = []
    pos = 0
    start = 0
    escaped = False
    skipping_ws = True
    head = ""
    chunks = iter(chunks)
    # The UNA service string advice is nine characters and may span chunks.
    for chunk in chunks:
        head += chunk
        if len(head) >= 9 or not head.startswith("UNA"[:len(head)]):
            break
    if head.startswith("UNA") and len(head) >= 9:
        pos = start = 9
        head = head[9:]

    def with_head():
        if head:
            yield head
        yield from chunks

    for chunk in with_head():
        i = 0
        n = len(chunk)
        while i < n:
            ch = chunk[i]
            if skipping_ws:
                if ch in "\r\n \t":
                    i += 1
                    pos += 1
                    start = pos
                    continue
                skipping_ws = False
            if escaped:
                buf.append(ch)
                escaped = False
            elif ch == release:
                buf.append(ch)
                escaped = True
            elif ch == terminator:
                seg = "".join(buf)
                if seg.strip():
                    yield start, seg
                buf = []
                skipping_ws = True
                i += 1
                pos += 1
                start = pos
                continue
            else:
                buf.append(ch)
            i += 1
            pos += 1
    tail = "".join(buf)
    if tail.strip():
        yield start, tail

File: src/lib/stream.ts
/*
 * Chunked readers for large files, the browser counterpart of
 * checkdigit/workspace/stream.py. A file is read in fixed slices and decoded
 * incrementally, so a multibyte sequence split across a slice boundary is never
 * corrupted; records are yielded with the absolute character offset (UTF-16
 * units) of their raw span, so a later pass can splice edits while streaming
 * the same file again.
 *
 * Readers:
 *   textChunks        bytes -> decoded text chunks with a chosen codec
 *   csvRecords        RFC 4180 records across quoted newlines and chunk boundaries
 *   lineRecords       one record per physical line
 *   delimitedSegments terminator-delimited segments honouring a release character
 *   xmlStartTags      start tags across chunk boundaries, skipping comments and CDATA
 *
 * The codec is strict UTF-8 unless a byte sequence fails to decode, in which
 * case the caller restarts the whole pass as ISO-8859-1 (the same rule as the
 * service). Latin-1 is decoded by hand: the browser's "latin1" label means
 * windows-1252, which would change bytes 0x80-0x9F.
 */

import type { Codec } from "./encoding";

export const CHUNK_BYTES = 4 * 1024 * 1024;

export class CodecError extends Error {
  constructor(public readonly byteOffset: number) {
    super(`not valid UTF-8 near byte ${byteOffset}`);
  }
}

export interface ByteSource {
  size: number;
  slice(start: number, end: number): Promise<Uint8Array>;
}

export function blobSource(blob: Blob): ByteSource {
  return {
    size: blob.size,
    slice: async (start, end) => new Uint8Array(await blob.slice(start, end).arrayBuffer()),
  };
}

export function bytesSource(bytes: Uint8Array): ByteSource {
  return { size: bytes.length, slice: async (start, end) => bytes.subarray(start, end) };
}

function latin1(bytes: Uint8Array): string {
  let text = "";
  const step = 8192;
  for (let i = 0; i < bytes.length; i += step) {
    text += String.fromCharCode(...bytes.subarray(i, Math.min(bytes.length, i + step)));
  }
  return text;
}

/** Decoded text chunks. Throws CodecError on invalid UTF-8 when codec is "utf-8". */
export async function* textChunks(source: ByteSource, codec: Codec = "utf-8", chunkBytes = CHUNK_BYTES,
  onBytes?: (done: number) => void): AsyncGenerator<string> {
  const decoder = codec === "utf-8" ? new TextDecoder("utf-8", { fatal: true, ignoreBOM: true }) : null;
  let offset = 0;
  while (offset < source.size) {
    const end = Math.min(source.size, offset + chunkBytes);
    const bytes = await source.slice(offset, end);
    let text: string;
    if (decoder) {
      try {
        text = decoder.decode(bytes, { stream: end < source.size });
      } catch {
        throw new CodecError(offset);
      }
    } else {
      text = latin1(bytes);
    }
    offset = end;
    onBytes?.(offset);
    if (text) yield text;
  }
}

export interface Field {
  col: number;
  value: string;
  start: number;
  end: number;
  quoted: boolean;
}

export interface Record {
  no: number;
  start: number;
  end: number;
  fields: Field[];
}

/** RFC 4180 state machine over a stream of text chunks; state survives chunk boundaries. */
export async function* csvRecords(chunks: AsyncIterable<string> | Iterable<string>, delimiter: string): AsyncGenerator<Record> {
  const special = new RegExp("[" + delimiter.replace(/[\\\]^-]/g, "\\$&") + '"\r\n]', "g");
  let row: Field[] = [];
  let buf: string[] = [];
  let inQuotes = false;
  let afterQuote = false;
  let quotedField = false;
  let fieldStart = 0;
  let recordStart = 0;
  let pos = 0;
  let recNo = 0;
  let pendingCr = false;
  let sawAny = false;

  for await (const chunk of chunks as AsyncIterable<string>) {
    let i = 0;
    const n = chunk.length;
    if (n) sawAny = true;
    if (pendingCr) {
      pendingCr = false;
      if (n && chunk[0] === "\n") {
        i = 1;
        pos += 1;
        fieldStart = pos;
        recordStart = pos;
      }
    }
    while (i < n) {
      if (inQuotes) {
        if (afterQuote) {
          afterQuote = false;
          if (chunk[i] === '"') {
            buf.push('"');
            i += 1;
            pos += 1;
            continue;
          }
          inQuotes = false;
        } else {
          const j = chunk.indexOf('"', i);
          if (j < 0) {
            buf.push(chunk.slice(i));
            pos += n - i;
            i = n;
            continue;
          }
          buf.push(chunk.slice(i, j));
          pos += j - i + 1;
          i = j + 1;
          afterQuote = true;
          continue;
        }
      }
      const ch = chunk[i] as string;
      if (ch === '"' && pos === fieldStart) {
        inQuotes = true;
        quotedField = true;
        i += 1;
        pos += 1;
        continue;
      }
      if (ch === delimiter) {
        row.push({ col: row.length, value: buf.join(""), start: fieldStart, end: pos, quoted: quotedField });
        buf = [];
        quotedField = false;
        i += 1;
        pos += 1;
        fieldStart = pos;
        continue;
      }
      if (ch === "\r" || ch === "\n") {
        row.push({ col: row.length, value: buf.join(""), start: fieldStart, end: pos, quoted: quotedField });
        buf = [];
        quotedField = false;
        const rec: Record = { no: recNo, start: recordStart, end: pos, fields: row };
        row = [];
        recNo += 1;
        let step = 1;
        if (ch === "\r") {
          if (i + 1 < n) step = chunk[i + 1] === "\n" ? 2 : 1;
          else pendingCr = true;
        }
        i += step;
        pos += step;
        fieldStart = pos;
        recordStart = pos;
        yield rec;
        continue;
      }
      special.lastIndex = i;
      const m = special.exec(chunk);
      const j = m ? m.index : n;
      buf.push(chunk.slice(i, j));
      pos += j - i;
      i = j;
    }
  }
  if (sawAny && (buf.length || row.length || quotedField || pos > recordStart)) {
    row.push({ col: row.length, value: buf.join(""), start: fieldStart, end: pos, quoted: quotedField });
    yield { no: recNo, start: recordStart, end: pos, fields: row };
  }
}

/** Yield [lineNo (1-based), absoluteStart, lineTextWithoutBreak]. */
export async function* lineRecords(chunks: AsyncIterable<string> | Iterable<string>): AsyncGenerator<[number, number, string]> {
  let carry = "";
  let pos = 0;
  let lineNo = 0;
  let pendingCr = false;
  for await (const chunk of chunks as AsyncIterable<string>) {
    const text = carry + chunk;
    carry = "";
    let start = 0;
    const n = text.length;
    if (pendingCr) {
      pendingCr = false;
      if (n && text[0] === "\n") start = 1;
    }
    let i = start;
    while (i < n) {
      const cr = text.indexOf("\r", i);
      const lf = text.indexOf("\n", i);
      let brk = -1;
      if (cr >= 0 && lf >= 0) brk = Math.min(cr, lf);
      else brk = cr >= 0 ? cr : lf;
      if (brk < 0) break;
      lineNo += 1;
      yield [lineNo, pos + start, text.slice(start, brk)];
      if (text[brk] === "\r") {
        if (brk + 1 < n) i = brk + (text[brk + 1] === "\n" ? 2 : 1);
        else {
          pendingCr = true;
          i = brk + 1;
        }
      } else {
        i = brk + 1;
      }
      start = i;
    }
    carry = text.slice(start);
    pos += start;
  }
  if (carry) {
    lineNo += 1;
    yield [lineNo, pos, carry];
  }
}

/**
 * Terminator-delimited segments (EDIFACT, X12). `release` is the escape character
 * ("?" for EDIFACT; "" when the syntax has none). The raw segment keeps release
 * characters so it can be spliced back verbatim. With `skipUna`, a nine-character
 * UNA service string advice is skipped even when it spans chunks.
 */
export async function* delimitedSegments(chunks: AsyncIterable<string> | Iterable<string>, terminator: string, release = "",
  skipUna = false): AsyncGenerator<[number, string]> {
  const iter = (chunks as AsyncIterable<string>)[Symbol.asyncIterator]
    ? (chunks as AsyncIterable<string>)[Symbol.asyncIterator]()
    : (async function* () { yield* chunks as Iterable<string>; })();
  let head = "";
  let pos = 0;
  let start = 0;
  if (skipUna) {
    while (head.length < 9 && head.startsWith("UNA".slice(0, head.length))) {
      const r = await iter.next();
      if (r.done) break;
      head += r.value;
      if (!head.startsWith("UNA".slice(0, Math.min(3, head.length)))) break;
    }
    if (head.startsWith("UNA") && head.length >= 9) {
      pos = start = 9;
      head = head.slice(9);
    }
  }
  const buf: string[] = [];
  let escaped = false;
  let skippingWs = true;
  const specialSrc = "[" + [terminator, release].filter(Boolean).map((c) => c.replace(/[\\\]^-]/g, "\\$&")).join("") + "]";
  const special = new RegExp(specialSrc, "g");

  async function* withHead(): AsyncGenerator<string> {
    if (head) yield head;
    while (true) {
      const r = await iter.next();
      if (r.done) return;
      yield r.value;
    }
  }
  for await (const chunk of withHead()) {
    let i = 0;
    const n = chunk.length;
    while (i < n) {
      const ch = chunk[i] as string;
      if (skippingWs) {
        if (ch === "\r" || ch === "\n" || ch === " " || ch === "\t") {
          i += 1;
          pos += 1;
          start = pos;
          continue;
        }
        skippingWs = false;
      }
      if (escaped) {
        buf.push(ch);
        escaped = false;
        i += 1;
        pos += 1;
        continue;
      }
      if (release && ch === release) {
        buf.push(ch);
        escaped = true;
        i += 1;
        pos += 1;
        continue;
      }
      if (ch === terminator) {
        const seg = buf.join("");
        if (seg.trim()) yield [start, seg];
        buf.length = 0;
        skippingWs = true;
        i += 1;
        pos += 1;
        start = pos;
        continue;
      }
      special.lastIndex = i;
      const m = special.exec(chunk);
      const j = m ? m.index : n;
      buf.push(chunk.slice(i, j));
      pos += j - i;
      i = j;
    }
  }
  const tail = buf.join("");
  if (tail.trim()) yield [start, tail];
}

export class XmlSecurityError extends Error {}

const MAX_XML_CONSTRUCT = 4 * 1024 * 1024;

/**
 * Start tags (including self-closing ones) as [absoluteOffset, tagText]. Comments,
 * CDATA sections, processing instructions and end tags are skipped; quotes inside
 * attribute values are honoured; any construct may span chunks. DOCTYPE and
 * ENTITY declarations raise XmlSecurityError. Well-formedness is not verified.
 */
export async function* xmlStartTags(chunks: AsyncIterable<string> | Iterable<string>): AsyncGenerator<[number, string]> {
  let carry = "";
  let base = 0;
  for await (const chunk of chunks as AsyncIterable<string>) {
    const text = carry + chunk;
    let i = 0;
    const n = text.length;
    while (true) {
      const lt = text.indexOf("<", i);
      if (lt < 0) {
        i = n;
        break;
      }
      if (n - lt < 9) {
        i = lt;
        break;
      }
      if (text.startsWith("<!--", lt)) {
        const end = text.indexOf("-->", lt + 4);
        if (end < 0) { i = lt; break; }
        i = end + 3;
        continue;
      }
      if (text.startsWith("<![CDATA[", lt)) {
        const end = text.indexOf("]]>", lt + 9);
        if (end < 0) { i = lt; break; }
        i = end + 3;
        continue;
      }
      if (text.startsWith("<?", lt)) {
        const end = text.indexOf("?>", lt + 2);
        if (end < 0) { i = lt; break; }
        i = end + 2;
        continue;
      }
      if (text.startsWith("<!", lt)) throw new XmlSecurityError("DOCTYPE or ENTITY declaration refused");
      let j = lt + 1;
      let quote = "";
      let end = -1;
      while (j < n) {
        const ch = text[j] as string;
        if (quote) {
          if (ch === quote) quote = "";
        } else if (ch === '"' || ch === "'") {
          quote = ch;
        } else if (ch === ">") {
          end = j;
          break;
        }
        j += 1;
      }
      if (end < 0) { i = lt; break; }
      if (text[lt + 1] !== "/") yield [base + lt, text.slice(lt, end + 1)];
      i = end + 1;
    }
    carry = text.slice(i);
    base += i;
    if (carry.length > MAX_XML_CONSTRUCT) throw new Error("XML construct longer than the streaming limit");
  }
  if (carry.trimStart().startsWith("<!")) throw new XmlSecurityError("DOCTYPE or ENTITY declaration refused");
}

/** Helper for tests and small inputs: split a string into fixed-size chunks. */
export function* splitChunks(text: string, size: number): Generator<string> {
  for (let i = 0; i < text.length; i += size) yield text.slice(i, i + size);
}
